### PR TITLE
kubectl: kubectl version better error message

### LINKED ISSUE #552: kubectl version: needs better error message when server not available
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

BUG REPORT

**Kubernetes version**:

v1.12.2

**Environment**:
- **Cloud provider or hardware configuration**: MacBook Pro 2017
- **OS**: MacOS High Sierra
- **Kernel**: Darwin Kernel Version 17.7.0
- **Install tools**: Homebrew
- **Others**: N/A

**What happened**:

When issuing the `kubectl version` command, in addition to the version information, an error was produced, as well as an output of some HTML.

```
~ $ kubectl version
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.2", GitCommit:"17c77c7898218073f14c8d573582e8d2313dc740", GitTreeState:"clean", BuildDate:"2018-10-30T21:39:38Z", GoVersion:"go1.11.1", Compiler:"gc", Platform:"darwin/amd64"}
error: got '<!DOCTYPE html>
<html lang="en" dir="ltr">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="robots" content="noindex">
<title>Login - Adminer</title>
<link rel="stylesheet" type="text/css" href="version?file=default.css&amp;version=4.6.3">
<script src='version?file=functions.js&amp;version=4.6.3' nonce="OTM5NDUyNzJhMzgzMmU1MjQ0MDc1MTJhMTIzOWZmYWE="></script>
<link rel="shortcut icon" type="image/x-icon" href="version?file=favicon.ico&amp;version=4.6.3">
<link rel="apple-touch-icon" href="version?file=favicon.ico&amp;version=4.6.3">

<body class="ltr nojs">
<script nonce="OTM5NDUyNzJhMzgzMmU1MjQ0MDc1MTJhMTIzOWZmYWE=">
mixin(document.body, {onkeydown: bodyKeydown, onclick: bodyClick, onload: partial(verifyVersion, '4.6.3', 'version?', '944539:431296')});
document.body.className = document.body.className.replace(/ nojs/, ' js');
var offlineMessage = 'You are offline.';
var thousandsSeparator = ',';
</script>

<div id="help" class="jush-sql jsonly hidden"></div>
<script nonce="OTM5NDUyNzJhMzgzMmU1MjQ0MDc1MTJhMTIzOWZmYWE=">mixin(qs('#help'), {onmouseover: function () { helpOpen = 1; }, onmouseout: helpMouseout});</script>

<div id="content">
<h2>Login</h2>
<div id='ajaxstatus' class='jsonly hidden'></div>
<form action='' method='post'>
<div></div>
<table cellspacing='0'>
<tr><th>System<td><select name='auth[driver]'><option value="server" selected>MySQL<option value="sqlite">SQLite 3<option value="sqlite2">SQLite 2<option value="pgsql">PostgreSQL<option value="oracle">Oracle (beta)<option value="mssql">MS SQL (beta)<option value="firebird">Firebird (alpha)<option value="simpledb">SimpleDB<option value="mongo">MongoDB<option value="elastic">Elasticsearch (beta)</select>
<tr><th>Server<td><input name="auth[server]" value="db" title="hostname[:port]" placeholder="localhost" autocapitalize="off">
<tr><th>Username<td><input name="auth[username]" id="username" value="" autocapitalize="off"><script nonce="OTM5NDUyNzJhMzgzMmU1MjQ0MDc1MTJhMTIzOWZmYWE=">focus(qs('#username'));</script>
<tr><th>Password<td><input type="password" name="auth[password]">
<tr><th>Database<td><input name="auth[db]" value="" autocapitalize="off">
</table>
<p><input type='submit' value='Login'>
<label><input type='checkbox' name='auth[permanent]' value='1'>Permanent login</label>
</form>
</div>

<form action='' method='post'>
<div id='lang'>Language: <select name='lang'><option value="en" selected>English<option value="ar">العربية<option value="bg">Български<option value="bn">বাংলা<option value="bs">Bosanski<option value="ca">Català<option value="cs">Čeština<option value="da">Dansk<option value="de">Deutsch<option value="el">Ελληνικά<option value="es">Español<option value="et">Eesti<option value="fa">فارسی<option value="fi">Suomi<option value="fr">Français<option value="gl">Galego<option value="he">עברית<option value="hu">Magyar<option value="id">Bahasa Indonesia<option value="it">Italiano<option value="ja">日本語<option value="ko">한국어<option value="lt">Lietuvių<option value="ms">Bahasa Melayu<option value="nl">Nederlands<option value="no">Norsk<option value="pl">Polski<option value="pt">Português<option value="pt-br">Português (Brazil)<option value="ro">Limba Română<option value="ru">Русский<option value="sk">Slovenčina<option value="sl">Slovenski<option value="sr">Српски<option value="ta">த‌மிழ்<option value="th">ภาษาไทย<option value="tr">Türkçe<option value="uk">Українська<option value="vi">Tiếng Việt<option value="zh">简体中文<option value="zh-tw">繁體中文</select><script nonce="OTM5NDUyNzJhMzgzMmU1MjQ0MDc1MTJhMTIzOWZmYWE=">qsl('select').onchange = function () { this.form.submit(); };</script> <input type='submit' value='Use' class='hidden'>
<input type='hidden' name='token' value='135968:716411'>
</div>
</form>
<div id="menu">
<h1>
<a href='https://www.adminer.org/' target="_blank" rel="noreferrer noopener" id='h1'>Adminer</a> <span class="version">4.6.3</span>
<a href="https://www.adminer.org/#download" target="_blank" rel="noreferrer noopener" id="version"></a>
</h1>
</div>
<script nonce="OTM5NDUyNzJhMzgzMmU1MjQ0MDc1MTJhMTIzOWZmYWE=">setupSubmitHighlight(document);</script>
': invalid character '<' looking for beginning of value
```

**What you expected to happen**:

Only version information to be printed.

**How to reproduce it**:

1. `brew install kubernetes-cli`
2. `kubectl version`

**Anything else we need to know**:

Nothing that I can think of, but feel free to ask questions and I'll be happy to help!

### PR DESCRIPTION
**Pull request by ahmadnurus**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Better error message when cluster server in kubeconfig is running but it is doesn't seem a kubernetes cluster server which causes an error when unmarshaling response body from the server into version.Info

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubectl/issues/552

**Special notes for your reviewer**:

/cc @soltysh @sttts 

This PR is a change from #71917

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/cmd/util/helpers.go [hunk 1]
```diff
@@ -142,6 +143,9 @@ func checkErr(err error, handleErr func(string, int)) {
 		}
 	case clientcmd.IsConfigurationInvalid(err):
 		handleErr(MultilineError("Error in configuration: ", err), DefaultErrorExitCode)
+	case discovery.IsParseServerVersionFailedError(err):
```
Comment (sttts) [inline pkg/kubectl/cmd/util/helpers.go]: why is the `default:` error handling below not enough?

Comment (prksu) [reply pkg/kubectl/cmd/util/helpers.go]: `default:` error handling below uses the type of error for comparison.

Comment (prksu) [reply pkg/kubectl/cmd/util/helpers.go]: default error handling which one you mean? there are two default there.

Comment (juanvallejo) [reply pkg/kubectl/cmd/util/helpers.go]: We're already matching the error's type in the nested switch below. You could just add a new `case *ErrParseServerVersionFailed:`. Also, if you're already wrapping the error in `&ErrParseServerVersionFailed{}`, and its `Error` method already prepends the message `unable to parse the server version`, I don't think it's necessary to state that again like you're doing in the line below.

Comment (prksu) [reply pkg/kubectl/cmd/util/helpers.go]: @juanvallejo do you mean by returning the error message `unable to parse the server version` is that enough?

Comment (prksu) [reply pkg/kubectl/cmd/util/helpers.go]: updated

**Review by soltysh — APPROVED** [review]
One nit, the rest lgtm.

File: pkg/kubectl/cmd/util/helpers.go [hunk 2]
```diff
@@ -155,6 +156,8 @@ func checkErr(err error, handleErr func(string, int)) {
 			default:
 				handleErr(fmt.Sprintf("the server doesn't have a resource type %q", err.PartialResource.Resource), DefaultErrorExitCode)
 			}
+		case *discovery.ErrParseServerVersionFailed:
+			handleErr(fmt.Sprintf("Error retrieving server version: %s", err.Error()), DefaultErrorExitCode)
```
[the remark this answers is not in the document]
Comment (prksu) [reply pkg/kubectl/cmd/util/helpers.go]: updated

File: pkg/kubectl/cmd/util/helpers.go [hunk 3]
```diff
@@ -155,6 +156,8 @@ func checkErr(err error, handleErr func(string, int)) {
 			default:
 				handleErr(fmt.Sprintf("the server doesn't have a resource type %q", err.PartialResource.Resource), DefaultErrorExitCode)
 			}
+		case *discovery.ErrParseServerVersionFailed:
+			handleErr(fmt.Sprintf("error retrieving server version: %s", err.Error()), DefaultErrorExitCode)
```
Comment (sttts) [inline pkg/kubectl/cmd/util/helpers.go]: Still the same question: why do we need special error handling in kubectl? We better report a good error message from the discovery code and use the default error path here.

File: staging/src/k8s.io/client-go/discovery/discovery_client.go [hunk 4]
```diff
@@ -381,7 +391,7 @@ func (d *DiscoveryClient) ServerVersion() (*version.Info, error) {
 	var info version.Info
 	err = json.Unmarshal(body, &info)
 	if err != nil {
-		return nil, fmt.Errorf("got '%s': %v", string(body), err)
+		return nil, &ErrParseServerVersionFailed{Err: err}
```
Comment (sttts) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: just `fmt.Errorf("unable to parse the server version: %s", err)`

File: pkg/kubectl/cmd/util/helpers.go (hunk 3 above)
Comment (prksu) [reply pkg/kubectl/cmd/util/helpers.go]: maybe we need different error message between client-go and kubectl. error message in kubectl giving the message what error and in client-go (discovery) giving a reason why it's an error.

of course, if this is not necessary i will update it.

File: staging/src/k8s.io/client-go/discovery/discovery_client.go (hunk 4 above)
Comment (prksu) [reply staging/src/k8s.io/client-go/discovery/discovery_client.go]: updated

**Pull request merged**